### PR TITLE
fix(install): Replace platform.dist with distro.linux_distribution for PY3.8+ compatibility

### PR DESCRIPTION
**Pull request by nikunj1222**
Replace platform.dist() with distro.linux_distribution in install.py | platform.dist() not supported by python3.8

Distro Information (Required)

Python >= 3.5
Ubuntu 20.04 LTS
And for all new distros in future

Possible Solution

To replace Replace platform.dist() with **distro.linux_distribution** in install.py

Use this module https://github.com/nir0s/distro for getting distro/system verion information.

Install distro module

`
Sudo pip install distro

`

### CODE REVIEW

File: install.py [hunk 1]
```diff
@@ -97,7 +97,11 @@ def check_distribution_compatibility():
 def get_distribution_info():
 	# return distribution name and major version
 	if platform.system() == "Linux":
-		current_dist = platform.dist()
+        if (platform.python_version_tuple()[0] + platform.python_version_tuple()[1]) == '2.7':
```
Comment (nikunj1222) [inline install.py:100]: If the it is python version 2.7 use  platform.dist() for python3 use dist.linux_distribution()

**Comment by nikunj1222** [comment]
The do think now the failure is coming from other dependencies...?

**Comment by nikunj1222** [comment]
#1019 https://github.com/frappe/bench/pull/1019

Should be merged to solved the errors on python 3.7,3.6,3.8

**Comment by nikunj1222** [comment]
#1016 #997 
Above issues can be solved with this pull request.

File: install.py [hunk 2]
```diff
@@ -10,7 +10,19 @@
...
+    import importlib
+    try:
+        importlib.import_module(package)
+    except ImportError:
+        import pip
+        pip.main(['install', package])
+    finally:
+        globals()[package] = importlib.import_module(package)
+
+install_and_import('distro')
```
Comment (Alchez) [inline install.py:25]: You can replace all of this with a simple `import distro`, and add that package in bench's `requirements.txt` file

File: install.py [hunk 3]
```diff
@@ -95,14 +106,23 @@ def check_distribution_compatibility():
...
-	elif platform.system() == "Darwin":
-		current_dist = platform.mac_ver()
-		return "macos", current_dist[0].rsplit('.', 1)[0]
+
+    # return distribution name and major version
+
+    if platform.system() == 'Linux':
+        if int(platform.python_version_tuple()[0]
+               + platform.python_version_tuple()[1]) <= 3.7:
+            current_dist = platfor.dist()
```
Comment (Alchez) [inline install.py:115]: You have a typo here

File: install.py [hunk 4]
```diff
@@ -95,14 +106,23 @@ def check_distribution_compatibility():
...
-
-	elif platform.system() == "Darwin":
-		current_dist = platform.mac_ver()
-		return "macos", current_dist[0].rsplit('.', 1)[0]
+
+    # return distribution name and major version
+
+    if platform.system() == 'Linux':
+        if int(platform.python_version_tuple()[0]
+               + platform.python_version_tuple()[1]) <= 3.7:
```
Comment (Alchez) [inline install.py:114]: You can use the `sys` module here

```suggestion
        if sys.version_info < (3, 7):
```

File: install.py [hunk 5]
```diff
@@ -95,14 +106,23 @@ def check_distribution_compatibility():
...
+            current_dist = platfor.dist()
+        else:
+            current_dist = \
+                distro.linux_distribution(full_distribution_name=True)
+
+            return (current_dist[0].lower(), current_dist[1].rsplit('.'
+                    )[0])
+    elif platform.system() == 'Darwin':
+
```
Comment (Alchez) [inline install.py:123]: Remove the extra line

File: install.py (hunk 2 above)
Comment (gavindsouza) [reply install.py:25]: @Alchez `install.py` should be able to set up a frappe env with a single command on a fresh server. It should manage all the dependencies it needs. Although they share the same repo, `install.py` isn't a part of bench, so we don't need to update `requirements.txt`. This should work fine.

File: install.py [hunk 6]
```diff
@@ -10,7 +10,19 @@
...
 import datetime
+import importlib
 
+def install_and_import(package):
+    import importlib
+    try:
+        importlib.import_module(package)
+    except ImportError:
+        import pip
+        pip.main(['install', package])
```
Comment (gavindsouza) [inline install.py:21]:
```suggestion
		subprocess.check_call('{0} -m pip install {1}'.format(sys.executable, package))
```
Start it as a subprocess instead. 

**References:**
1. https://pip.pypa.io/en/stable/user_guide/?highlight=_internal#using-pip-from-your-program
2. https://stackoverflow.com/a/47059613/10309266

File: install.py [hunk 7]
```diff
@@ -95,14 +106,23 @@ def check_distribution_compatibility():
...
+    if platform.system() == 'Linux':
+        if int(platform.python_version_tuple()[0]
+               + platform.python_version_tuple()[1]) <= 3.7:
+            current_dist = platfor.dist()
+        else:
+            current_dist = \
+                distro.linux_distribution(full_distribution_name=True)
+
+            return (current_dist[0].lower(), current_dist[1].rsplit('.'
+                    )[0])
```
Comment (Alchez) [inline install.py:121]: You can use some other [`distro` functions](https://distro.readthedocs.io/en/latest/#consolidated-accessor-functions) for what you're trying to do:

```suggestion
            return (distro.id(), distro.major_version())
```

File: install.py (hunk 2 above)
Comment (Alchez) [reply install.py:25]: Will this need to be done again in case there's another dependency to be added?

File: install.py [hunk 8]
```diff
@@ -10,7 +10,18 @@
...
 import warnings
 import datetime
+import importlib
 
+def install_and_import(package):
+    import importlib
+    try:
+        importlib.import_module(package)
+    except ImportError:
+		subprocess.check_call('{0} -m pip install {1}'.format(sys.executable, package))
```
Comment (gavindsouza) [inline install.py:20]: I think `check_call` takes in an iterable for a command, like 
```python
>>> subprocess.check_call([sys.executable, "-m", "pip", "install", package]) 
```

File: install.py [hunk 9]
```diff
@@ -10,7 +10,18 @@
 import platform
 import warnings
 import datetime
+import importlib
 
+def install_and_import(package):
+    import importlib
```
Comment (gavindsouza) [inline install.py:16]:
```suggestion
```

**Comment by nikunj1222** [comment]
Hi guys it seems all checks are passed. Let me know if you need any more changes..

File: install.py [hunk 10]
```diff
@@ -10,7 +10,18 @@
 import platform
 import warnings
 import datetime
+import importlib
 
+def install_and_import(package):
```
Comment (gavindsouza) [inline install.py:15]: * Rename it to `install_python_package` instead, that'd fit with the convention and is more intuitive. 
* Move this function down while maintaining 2 line spacing with the other functions

File: install.py [hunk 11]
```diff
@@ -10,7 +10,18 @@
...
+    try:
+        importlib.import_module(package)
+    except ImportError:
+	print("Trying to Install required module: " + package)
+	subprocess.check_call([sys.executable, "-m", "pip", "install", package])
+    finally:
+        globals()[package] = importlib.import_module(package)
+
+install_and_import('distro')
```
Comment (gavindsouza) [inline install.py:24]: This should be called as the last step in the `install_prerequisites` function

File: install.py [hunk 12]
```diff
@@ -95,14 +105,16 @@ def check_distribution_compatibility():
...
+
+    # return distribution name and major version
+
+	if platform.system() == 'Linux':
+            return (distro.id(), distro.major_version())
+	
+        elif platform.system() == 'Darwin':
+
+            current_dist = platform.mac_ver()
+            return ('macos', current_dist[0].rsplit('.', 1)[0])
```
Comment (gavindsouza) [inline install.py:117]: Indentations are off here. Use tabs only and remove unnecessary new-lines on L108, L110 and L115

File: install.py [hunk 13]
```diff
@@ -10,7 +10,18 @@
...
+import importlib
 
+def install_and_import(package):
+    try:
+        importlib.import_module(package)
+    except ImportError:
+	print("Trying to Install required module: " + package)
+	subprocess.check_call([sys.executable, "-m", "pip", "install", package])
+    finally:
+        globals()[package] = importlib.import_module(package)
```
Comment (gavindsouza) [inline install.py:22]: Inconsistent indentation!

File: install.py [hunk 14]
```diff
@@ -433,6 +445,7 @@ def parse_commandline_args():
 	with warnings.catch_warnings():
 		warnings.simplefilter("ignore")
 		setup_log_stream(args)
+		install_python_package('distro')
```
Comment (gavindsouza) [inline install.py:448]: this breaks if pip isn't already installed on the system. Specifically for Python versions <= PY2.7.9 and <= PY3.4

pip is installed later on in `install_prerequisites` as of right now.

Comment (nikunj1222) [reply install.py:448]: @gavindsouza  How shall this be resolved then. 
we will not install the prerequisite if the system is not compatible. OR we install prerequisite before checking the OS.

Comment (gavindsouza) [reply install.py]: By installing pip first. I've pushed the changes to this branch.

**Comment by gavindsouza** [comment]
@nikunj1222 sorry we couldn't get this fixed earlier. but https://github.com/frappe/bench/pull/1048 has been merged. Thanks for your contribution.

**Pull request closed without merging**